RNA-seq/06-openpbta_heatmap.Rmd

```diff
@@ -0,0 +1,385 @@
+histology_colors <- unname(histology_colors)
+
+# annotation color list for ComplexHeatMap
+sample_annotation_colors <- list(
+  short_histology = c(
+    "LGAT" = histology_colors[[1]],
+    "Ependymoma" = histology_colors[[2]],
+    "HGAT" = histology_colors[[3]],
+    "Medulloblastoma" = histology_colors[[4]]
+  ),
+  composition = c(
+    "Solid Tissue" = "#A0A0A0",  # light grey
+    "Derived Cell Line" = "#000000"  # black
+  ),
+  tumor_descriptor = c(
+    "Initial CNS Tumor" = "#3333FF",
+    "Progressive" = "#FFFF99",
+    "Recurrence" = "#CCCCFF",
+    "Second Malignancy" = "#000033",
+    "Unavailable" = "#FFFFFF"  # white for missing data
+  )
+)
+
+# Customize the legend titles
+legend_titles =  list(
+    short_histology = list(title = "Histology"),
+    composition = list(title = "Composition"),
+    tumor_descriptor = list(title = "Tumor descriptor")
+    )
+```
+
+We need to create a special type of object with `HeatmapAnnotation` using the annotation data frame and the list of color vectors.
+
+```{r create_annotation, live = TRUE}
+column_annotation <- HeatmapAnnotation(
+  df = sample_annotation_df,
+  col = sample_annotation_colors,
+  annotation_label = c("Histology", "Composition", "Tumor descriptor")
+)
+```
+
+### Values for display
+
+We will z-score the expression values for display.
+This is sometimes called a standardized score.
+Some heatmap plotting packages will do this for you, but `ComplexHeatmap` does not.
+It's calculated for each value in a row (gene) by subtracting the gene's mean and dividing by the gene's standard deviation.
+This will result in every row having a mean of 0 and a standard deviation of 1.
+
+```{r zscore}
+zscores_mat <- 
+  (high_var_mat - rowMeans(high_var_mat)) / matrixStats::rowSds(high_var_mat)
+```
+
+Since all genes have the same z-score variance (by definition!), if you're filtering to high variance values, it's important to do that _prior_ to standardizing.
+
+### Heatmap itself!
+
+Okay, now we're ready to make a heatmap complete with annotation bars.
+
+```{r make_heatmap}
+Heatmap(zscores_mat,
+        # The distance metric used for clustering the rows
+        # This is different from the default (Euclidean)
+        clustering_distance_rows = "pearson",
+        # Linkage method for row clustering
+        # This is different from the default (complete)
+        clustering_method_rows = "average",
+        # Distance metric for columns
+        clustering_distance_columns = "pearson",
+        # Linkage for columns
+        clustering_method_columns = "average",
+        show_row_names = FALSE,
+        show_column_names = FALSE,
+        # Add annotation bars to the top of the 
+        top_annotation = column_annotation,
+        # This will be used as the label for the color bar
+        # of the cells of the heatmap itself
+        name = "z-score")
+```
+
+We have to create the heatmap _twice_ -- once for display in the notebook and once to save to a PNG file.
```

Comment:
Is this still true? I can never keep up.

Reply:
I tried to find a way not to do this, but didn't. 

I also had a twitter conversation with Jokergoo about a regression in the current version where you can't customize rasterizing. He implemented "better" rasterization without thinking it through for all cases, so now you get good screen resolution rasterization, but can't push resolution higher for output except by turning off rasterization... The net result is that I am actually doing something a bit different for the output block (`use_raster = FALSE`).